### PR TITLE
EOS-26836: wait function in init is not reliable

### PR DESCRIPTION
**Pull request by mssawant**
It is possible that not all the remote and local process statuses
are updated on every node. For more reliability it is better to
check all the processes status per node explicity.

Solution:
Iterate over each node's processes and check the corresponding status
in Consul KV.

### CODE REVIEW

**Comment by mssawant** [comment]
Single node provisioning completed fine,
```
[root@ssc-vm-1623:root] /opt/seagate/cortx/hare/bin/hare_setup init --config json:///root/hare.config.conf.1-node
Cluster is not running
2021-12-07 17:22:16: Generating cluster configuration... OK
2021-12-07 17:22:17: Stopping Consul on this node... OK
2021-12-07 17:22:29: Stopping Consul on other nodes... OK
2021-12-07 17:22:29: Starting Consul server on this node............ OK
2021-12-07 17:22:39: Importing configuration into the KV store... OK
2021-12-07 17:22:39: Starting Consul on other nodes...Consul ready on all nodes
2021-12-07 17:22:39: Updating Consul configuraton from the KV store... OK
2021-12-07 17:22:40: Waiting for the RC Leader to get elected....... OK
2021-12-07 17:22:45: Starting Motr (phase1, mkfs)... OK
2021-12-07 17:22:51: Starting Motr (phase1, m0d)... OK
2021-12-07 17:22:52: Starting Motr (phase2, mkfs)... OK
2021-12-07 17:23:03: Starting Motr (phase2, m0d)... OK
2021-12-07 17:23:07: Starting S3 servers (phase3)... OK
2021-12-07 17:23:08: Checking health of services... OK
2021-12-07 17:23:09,330 [INFO] processes {'ssc-vm-1623.colo.seagate.com': [0x7200000000000001:0xa, 0x7200000000000001:0x1b, 0x7200000000000001:0x2c, 0x7200000000000001:0x2f, 0x7200000000000001:0x7]}
2021-12-07 17:23:09,340 [INFO] Waiting for all the processes to start..
Stopping s3server@0x7200000000000001:0x2f at ssc-vm-1623.colo.seagate.com...
Stopped s3server@0x7200000000000001:0x2f at ssc-vm-1623.colo.seagate.com
Stopping m0d@0x7200000000000001:0xa (ios) at ssc-vm-1623.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x1b (ios) at ssc-vm-1623.colo.seagate.com...
Stopped m0d@0x7200000000000001:0xa (ios) at ssc-vm-1623.colo.seagate.com
Stopped m0d@0x7200000000000001:0x1b (ios) at ssc-vm-1623.colo.seagate.com
Stopping m0d@0x7200000000000001:0x2c (confd) at ssc-vm-1623.colo.seagate.com...
Stopped m0d@0x7200000000000001:0x2c (confd) at ssc-vm-1623.colo.seagate.com
Stopping hare-hax at ssc-vm-1623.colo.seagate.com...
Stopped hare-hax at ssc-vm-1623.colo.seagate.com
Making sure that RC leader can be re-elected next time
Cluster is not running
[root@ssc-vm-1623:root]
```

File: hax/hax/util.py [hunk 1]
```diff
@@ -820,6 +835,16 @@ def get_node_ctrl_fids(self,
             list_fids.append(Fid.parse(ctrl_fid))
         return list_fids
 
+    def get_node_hare_motr_s3_fids(self, node: str) -> List[Fid]:
```
Comment (knekrasov) [inline hax/hax/util.py:838]: Docstring is highly desirable given it is a utility function (hence it is intended to be reused in the future)

File: hax/hax/util.py [hunk 2]
```diff
@@ -820,6 +835,16 @@ def get_node_ctrl_fids(self,
...
+        services = ['hax', 'ios', 'confd', 's3service']
+        node_data: List[Dict[str, Any]] = self.get_node_health_details(node)
+        fids: List[Fid] = []
+        if node_data:
+            for item in node_data:
+                if item['ServiceName'] in services:
+                    fids.append(mk_fid(ObjT.PROCESS, int(item['ServiceID'])))
```
Comment (knekrasov) [inline hax/hax/util.py:845]: [Optional] This version is more readable:
```suggestion
        if not node_data:
            return []
 
        for item in node_data:
            if item['ServiceName'] in services:
                fids.append(mk_fid(ObjT.PROCESS, int(item['ServiceID'])))

```

**Review by knekrasov — COMMENTED** [review]
@mssawant Can you please elaborate a bit in which case the old version of the algorithm would be unreliable? In other words, I'd like to understand better the problem you are solving. Thank you.

**Review by vaibhavparatwar — COMMENTED** [review]
http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-3Node-CentOS-7.9/375/ 
Deployment for 3N VM from CICD was passed @mssawant  @knekrasov

**Comment by mssawant** [comment]
@knekrasov, sure. Hare maintains local as well remote processes statuses per node. In order to check if the cluster is up, instead of iterating over all the nodes and its services and their statuses, we were just counting the the total number of motr and s3services that are supposed to be present in cluster and then fetching the statuses for all the processes on a given node. From the fetched statuses we count the number of processes in STARTED state and match the count with the number of services that are supposed to STARTED. This count was mismatching because one of remote s3server status was not accounted in every node's processes statuses,
```
[root@ssc-vm-rhev4-0956 ~]# hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x7e 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xd4 'Profile_the_pool': 'storage-set-1__sns' None None
Services:
    srvnode-2.data.private  (RC)
    [started]  hax        0x7200000000000001:0x30  192.168.90.13@tcp:12345:1:1
    [started]  ioservice  0x7200000000000001:0x33  192.168.90.13@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0x3e  192.168.90.13@tcp:12345:2:2
    [started]  confd      0x7200000000000001:0x49  192.168.90.13@tcp:12345:2:3
    [started]  s3server   0x7200000000000001:0x4c  192.168.90.13@tcp:12345:3:1
    [unknown]  m0_client  0x7200000000000001:0x4f  192.168.90.13@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x52  192.168.90.13@tcp:12345:4:2
    srvnode-1.data.private
    [started]  hax        0x7200000000000001:0x7   192.168.85.73@tcp:12345:1:1
    [started]  ioservice  0x7200000000000001:0xa   192.168.85.73@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0x15  192.168.85.73@tcp:12345:2:2
    [started]  confd      0x7200000000000001:0x20  192.168.85.73@tcp:12345:2:3
    [started]  s3server   0x7200000000000001:0x23  192.168.85.73@tcp:12345:3:1
    [unknown]  m0_client  0x7200000000000001:0x26  192.168.85.73@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x29  192.168.85.73@tcp:12345:4:2
    srvnode-3.data.private
    [started]  hax        0x7200000000000001:0x59  192.168.90.187@tcp:12345:1:1
    [started]  ioservice  0x7200000000000001:0x5c  192.168.90.187@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0x67  192.168.90.187@tcp:12345:2:2
    [started]  confd      0x7200000000000001:0x72  192.168.90.187@tcp:12345:2:3
    [started]  s3server   0x7200000000000001:0x75  192.168.90.187@tcp:12345:3:1
    [unknown]  m0_client  0x7200000000000001:0x78  192.168.90.187@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x7b  192.168.90.187@tcp:12345:4:2
[root@ssc-vm-rhev4-0956 ~]# consul kv get -recurse srvnode-2.data.private
srvnode-2.data.private/processes/0x7200000000000001:0x15:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0x20:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0x30:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0x33:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0x3e:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0x49:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0x4c:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_OTHER"}
srvnode-2.data.private/processes/0x7200000000000001:0x5c:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0x67:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0x72:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-2.data.private/processes/0x7200000000000001:0xa:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
[root@ssc-vm-rhev4-0956 ~]# consul kv get -recurse srvnode-2.data.private | wc -l
11
[root@ssc-vm-rhev4-0956 ~]# consul kv get -recurse srvnode-3.data.private | wc -l
11
[root@ssc-vm-rhev4-0956 ~]# consul kv get -recurse srvnode-3.data.private
srvnode-3.data.private/processes/0x7200000000000001:0x15:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x20:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x33:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x3e:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x49:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x59:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x5c:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x67:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x72:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
srvnode-3.data.private/processes/0x7200000000000001:0x75:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_OTHER"}
srvnode-3.data.private/processes/0x7200000000000001:0xa:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
```
As you see above, total number of services to be running must be 12 (excluding hax, we count hax separately just once) but the number of statuses updated in consul are just 11.
This may happen due to following reasons,
In main, consul watcher uses systemd to watch motr/s3server processes. For remote processes, we rely on consul notification and accordingly update consul kv. It may happen that consul reported online for a remote s3server before it already started (from motr perspective), when hare receives this notification it checks for the current status of the process, if the process is not ready hare reports `UNKNOWN` for this process and does not update consul kv. Later if remote process started, peer node's will not get another online notification from Consul as it was already reported online. Thus, it may happen that the process's status is updated for its local node (as seen above) but might get missed on the remote nodes.
Thus the mechanism to check the cluster status is not reliable in such situations.

File: hax/hax/util.py (hunk 1 above)
Comment (mssawant) [reply hax/hax/util.py:838]: Done.

**Review by vaibhavparatwar — APPROVED** [review]

File: hax/hax/util.py (hunk 2 above)
Comment (mssawant) [reply hax/hax/util.py]: Okay, done.

**Review by knekrasov — APPROVED** [review]
Got it. Effectively, Consul KV can store some outdated keys and values that were left after re-bootstrapping or something. Makes sense, thank you.

**Pull request merged**